File: shell/components/gnome-shell-extension/lib/commandSurface.js
// SPDX-FileCopyrightText: 2026 ComradeArt
// SPDX-License-Identifier: GPL-3.0-or-later
//
// Universal command surface: Super+Space, character click, or the search
// field. Short answers stay in the companion bubble. Longer work opens a
// task card. This is not a chatbot transcript.

import {buildBubble, buildSearchField, buildTaskCard, limitBubbleText} from './design/primitives.js';
import {SCREEN_QUESTIONS} from './design/tokens.js';

/** How the surface is opened. Keyboard remains sufficient without the figure. */
export const COMMAND_TRIGGERS = Object.freeze([
    'super-space', 'character-click', 'search-entry',
]);

/** Super+Space in the GSettings schema (`open-launcher`). */
export const COMMAND_ACCELERATOR = '<Super>space';

const BUBBLE_CHAR_LIMIT = 220;

function sentenceCount(value) {
    const raw = String(value ?? '').replace(/\s+/g, ' ').trim();
    if (!raw)
        return 0;
    return raw.split(/(?<=[.!?])\s+/).filter(Boolean).length;
}

/**
 * Where an answer should land. Captions stay in the bubble; work does not
 * accumulate as a chat log.
 */
export function routeCommandAnswer(text, {
    working = false, stages = [], title = '',
} = {}) {
    const original = String(text ?? '').replace(/\s+/g, ' ').trim();
    const bubbleText = limitBubbleText(original);
    const namedStages = Array.isArray(stages)
        ? stages.filter(item => typeof item === 'string' && item.trim())
        : [];
    const long = Boolean(working)
        || namedStages.length > 0
        || sentenceCount(original) > 3
        || original.length > BUBBLE_CHAR_LIMIT;
    const surface = long ? 'task-card' : 'bubble';
    return {
        surface,
        bubble: buildBubble({text: bubbleText || original}),
        taskCard: surface === 'task-card'
            ? buildTaskCard({
                title: title || 'Working',
                caption: bubbleText || original,
                stages: namedStages.length ? namedStages : ['Plan', 'Run', 'Done'],
                stageIndex: working ? 1 : 2,
            })
            : null,
        transcript: false,
        companionRequired: false,
    };
}

/**
 * The empty command surface: a search field, not a thread.
 *
 * `companionRequired` is always false: Super+Space and the search field still
 * work when the figure is hidden.
 */
export function buildCommandSurface({
    trigger = 'search-entry', query = '', open = true,
} = {}) {
    const resolved = COMMAND_TRIGGERS.includes(trigger) ? trigger : 'search-entry';
    return {
        kind: 'CommandSurface',
        trigger: resolved,
        accelerator: COMMAND_ACCELERATOR,
        query: String(query ?? ''),
        open: Boolean(open),
        field: buildSearchField({
            value: query,
            placeholder: 'Search or ask Bunny',
        }),
        accessibleName: 'Search or ask Bunny',
        companionRequired: false,
        transcript: false,
        chatbot: false,
        opens: 'search-field',
        questions: SCREEN_QUESTIONS,
        next: 'Type a search or a request. Escape dismisses.',
    };
}

File: shell/services/bunny_shell/command_surface.py
# SPDX-FileCopyrightText: 2026 ComradeArt
# SPDX-License-Identifier: GPL-3.0-or-later
"""Universal command surface: Super+Space, character click, search entry.

Short answers stay in the companion bubble. Longer work opens a task card.
This module is the GTK/host twin of ``lib/commandSurface.js``.
"""

from __future__ import annotations

from dataclasses import dataclass

COMMAND_TRIGGERS = ("super-space", "character-click", "search-entry")
COMMAND_ACCELERATOR = "<Super>space"
_BUBBLE_CHAR_LIMIT = 220


def _sentence_count(value: str) -> int:
    raw = " ".join(str(value or "").split()).strip()
    if not raw:
        return 0
    parts: list[str] = []
    buf = []
    for char in raw:
        buf.append(char)
        if char in ".!?":
            parts.append("".join(buf).strip())
            buf = []
    if buf and "".join(buf).strip():
        parts.append("".join(buf).strip())
    return len(parts)


def _limit_bubble(value: str, max_sentences: int = 3) -> str:
    raw = " ".join(str(value or "").split()).strip()
    if not raw:
        return ""
    parts: list[str] = []
    buf: list[str] = []
    for char in raw:
        buf.append(char)
        if char in ".!?":
            parts.append("".join(buf).strip())
            buf = []
    if buf and "".join(buf).strip():
        parts.append("".join(buf).strip())
    if len(parts) <= max_sentences:
        return raw
    return " ".join(parts[:max_sentences])


@dataclass(frozen=True)
class CommandRoute:
    surface: str
    bubble_text: str
    task_title: str | None
    transcript: bool = False
    companion_required: bool = False


def route_command_answer(
    text: str,
    *,
    working: bool = False,
    stages: tuple[str, ...] = (),
    title: str = "",
) -> CommandRoute:
    original = " ".join(str(text or "").split()).strip()
    bubble = _limit_bubble(original)
    named = tuple(item for item in stages if str(item).strip())
    long = bool(working) or bool(named) or _sentence_count(original) > 3 or len(original) > _BUBBLE_CHAR_LIMIT
    return CommandRoute(
        surface="task-card" if long else "bubble",
        bubble_text=bubble or original,
        task_title=(title or "Working") if long else None,
        transcript=False,
        companion_required=False,
    )


@dataclass(frozen=True)
class CommandSurface:
    trigger: str
    query: str
    accelerator: str = COMMAND_ACCELERATOR
    companion_required: bool = False
    transcript: bool = False
    chatbot: bool = False
    accessible_name: str = "Search or ask Bunny"


def build_command_surface(*, trigger: str = "search-entry", query: str = "") -> CommandSurface:
    resolved = trigger if trigger in COMMAND_TRIGGERS else "search-entry"
    return CommandSurface(trigger=resolved, query=str(query or ""))
